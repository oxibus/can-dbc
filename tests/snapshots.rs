#![cfg(all(feature = "serde", feature = "encodings"))]

use nom::Err;
use std::borrow::Cow;
use std::fs;
use std::path::{Path, PathBuf};

use can_dbc::{decode_cp1252, Dbc, Error};
use insta::{assert_debug_snapshot, assert_yaml_snapshot, with_settings};
use test_each_file::test_each_path;

struct TestConfig {
    test_root: &'static str,
    snapshot_suffix: &'static str,
    use_cp1251: bool,
    create_snapshot: bool,
}

static TEST_DIRS: &[TestConfig] = &[
    TestConfig {
        test_root: "shared-test-files",
        snapshot_suffix: "",
        use_cp1251: true,
        create_snapshot: true,
    },
    TestConfig {
        test_root: "opendbc/opendbc/dbc",
        snapshot_suffix: "opendbc",
        use_cp1251: false,
        create_snapshot: false,
    },
];

test_each_path! { for ["dbc"] in "./tests/fixtures/opendbc/opendbc/dbc" as dbc => parse_one_file }
test_each_path! { for ["dbc"] in "./tests/fixtures/shared-test-files" as shared => parse_one_file }
// upper case extension
test_each_path! { for ["DBC"] in "./tests/fixtures/shared-test-files" as shared2 => parse_one_file }

/// Get `test root`, `snapshot name suffix`, `use cp1251`, `create snapshot` for the given path
fn get_test_info(path: &Path) -> Option<(PathBuf, &'static TestConfig)> {
    if !path
        .extension()
        .unwrap_or_default()
        .eq_ignore_ascii_case("dbc")
    {
        return None;
    }
    let path_str = path.to_str().unwrap();
    let parent = path.parent().unwrap();
    for item in TEST_DIRS {
        // Ensure slashes are there for easier matching
        let test_root = format!("/{}/", item.test_root);
        let mut path_dir = parent.to_str().unwrap().to_string();
        if !path_dir.ends_with('/') {
            path_dir.push('/');
        }
        if let Some(pos) = path_dir.find(&test_root) {
            let parent = PathBuf::from("snapshots")
                .join(item.snapshot_suffix)
                .join(&path_dir[pos + test_root.len()..]);
            return Some((parent, item));
        }
    }
    panic!("Unknown test directory: {path_str}");
}

/// Test parsing all DBC files
#[test]
fn test_if_submodules_are_present() {
    for test in TEST_DIRS {
        let dir = Path::new("./tests/fixtures").join(test.test_root);
        fs::read_dir(&dir).unwrap_or_else(|e| {
            let dir_display = dir.display();
            panic!(
                "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
            )
        });
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file([path]: [&Path; 1]) {
    let Some((
        snapshot_path,
        &TestConfig {
            use_cp1251,
            create_snapshot,
            ..
        },
    )) = get_test_info(path)
    else {
        return;
    };

    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    let buffer = if use_cp1251 {
        decode_cp1252(&buffer)
            .unwrap_or_else(|| panic!("Failed to decode {} as cp1252", path.display()))
    } else {
        Cow::Borrowed(
            std::str::from_utf8(&buffer)
                .unwrap_or_else(|_| panic!("Failed to decode {} as utf-8", path.display())),
        )
    };

    report_test_result(
        Dbc::try_from(buffer.as_ref()),
        snapshot_path,
        file_name,
        create_snapshot,
    );
}

fn report_test_result(
    result: Result<Dbc, Error>,
    snapshot_path: PathBuf,
    file_name: String,
    do_snapshot: bool,
) {
    let do_snapshot = std::env::var("SKIP_INSTA").is_err()
        && (do_snapshot || std::env::var("FORCE_INSTA").is_ok());

    with_settings! {
        {
            omit_expression => true,
            snapshot_path => snapshot_path,
            prepend_module_to_snapshot => false
        },
        {
            match result {
                Ok(result) => {
                    if do_snapshot {
                        assert_yaml_snapshot!(file_name, result);
                    } // else ignore successful results when not snapshotting
                }
                Err(e) => {
                    let file_name = format!("!error___{file_name}");
                    assert_debug_snapshot!(file_name, err_to_str(&e));
                }
            }
        }
    }
}

fn err_to_str<'a>(err: &'a Error) -> &'a str {
    match err {
        Error::Incomplete(_, _) => "incomplete",
        Error::Nom(e) => match e {
            Err::Incomplete(_) => "nom-incomplete",
            Err::Error(_) => "nom-error",
            Err::Failure(_) => "nom-failure",
        },
        Error::MultipleMultiplexors => "multiple multiplexors",
    }
}
