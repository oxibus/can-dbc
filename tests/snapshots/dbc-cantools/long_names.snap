---
source: tests/snapshots.rs
---
version: ""
new_symbols:
  - NS_DESC_
  - CM_
  - BA_DEF_
  - BA_
  - VAL_
  - CAT_DEF_
  - CAT_
  - FILTER
  - BA_DEF_DEF_
  - EV_DATA_
  - ENVVAR_DATA_
  - SGTYPE_
  - SGTYPE_VAL_
  - BA_DEF_SGTYPE_
  - BA_SGTYPE_
  - SIG_TYPE_REF_
  - VAL_TABLE_
  - SIG_GROUP_
  - SIG_VALTYPE_
  - SIGTYPE_VALTYPE_
  - BO_TX_BU_
  - BA_DEF_REL_
  - BA_REL_
  - BA_DEF_DEF_REL_
  - BU_SG_REL_
  - BU_EV_REL_
  - BU_BO_REL_
  - SG_MUL_VAL_
bit_timing: []
nodes:
  - NN123456789012345678901234567890
  - N12345678901234567890123456_0000
  - N1234567890123456789012345678901
  - N12345678901234567890123456_0001
messages:
  - id:
      Standard: 9
    name: SS123456789012345678901234587890
    size: 8
    transmitter: VectorXXX
  - id:
      Standard: 8
    name: SS1234567890123456789012345_0000
    size: 8
    transmitter: VectorXXX
  - id:
      Standard: 7
    name: SS1234567890123456789012345_0001
    size: 8
    transmitter: VectorXXX
  - id:
      Standard: 6
    name: SS123456789012345678901234577890
    size: 8
    transmitter: VectorXXX
  - id:
      Standard: 5
    name: SS123456789012345678901234567890
    size: 8
    transmitter: VectorXXX
  - id:
      Standard: 4
    name: S1234567890123456789012345678901
    size: 8
    transmitter: VectorXXX
    signals:
      - name: SS123456789012345678901234567890
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - id:
      Standard: 0
    name: M12345678901234567890123456_0000
    size: 8
    transmitter: VectorXXX
    signals:
      - name: SSS12345678901234567890123456789
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - id:
      Standard: 1
    name: M1234567890123456789012345678901
    size: 8
    transmitter:
      NodeName: N1234567890123456789012345678901
    signals:
      - name: SS1234567890123456789012345_0000
        multiplexer_indicator: Plain
        start_bit: 32
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: SS1234567890123456789012345_0001
        multiplexer_indicator: Plain
        start_bit: 24
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: SS1234567890123456789012345_0002
        multiplexer_indicator: Plain
        start_bit: 16
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - N12345678901234567890123456_0000
      - name: S12345678901234567890123456_0000
        multiplexer_indicator: Plain
        start_bit: 8
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: S1234567890123456789012345678901
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - id:
      Standard: 2
    name: M12345678901234567890123456_0001
    size: 8
    transmitter:
      NodeName: N12345678901234567890123456_0001
    signals:
      - name: SS1234567890123456789012345_0003
        multiplexer_indicator: Plain
        start_bit: 24
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: SS1234567890123456789012345_0004
        multiplexer_indicator: Plain
        start_bit: 16
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: S12345678901234567890123456_0001
        multiplexer_indicator: Plain
        start_bit: 8
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: S12345678901234567890123456_0002
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - id:
      Standard: 3
    name: MM123456789012345678901234567890
    size: 8
    transmitter: VectorXXX
    signals:
      - name: SSS123456789012345678901234_0000
        multiplexer_indicator: Plain
        start_bit: 8
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: SS1234567890123456789012345_0005
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
environment_variables:
  - name: E12345678901234567890123456_0000
    typ: Integer
    min: 0
    max: 0
    unit: ""
    initial_value: 0
    ev_id: 1
    access_type: DummyNodeVector0
    access_nodes:
      - Name: Vector__XXX
  - name: E1234567890123456789012345678901
    typ: Integer
    min: 0
    max: 0
    unit: ""
    initial_value: 0
    ev_id: 2
    access_type: DummyNodeVector0
    access_nodes:
      - Name: Vector__XXX
attribute_definitions:
  - Plain: "\"BusType\" STRING"
  - Signal: "\"SystemSignalLongSymbol\" STRING"
  - Message: "\"SystemMessageLongSymbol\" STRING"
  - Node: "\"SystemNodeLongSymbol\" STRING"
  - EnvironmentVariable: "\"SystemEnvVarLongSymbol\" STRING"
attribute_defaults:
  - name: BusType
    value:
      String: ""
  - name: SystemSignalLongSymbol
    value:
      String: ""
  - name: SystemMessageLongSymbol
    value:
      String: ""
  - name: SystemNodeLongSymbol
    value:
      String: ""
  - name: SystemEnvVarLongSymbol
    value:
      String: ""
attribute_values_node:
  - name: SystemNodeLongSymbol
    node_name: NN123456789012345678901234567890
    value:
      String: NN123456789012345678901234567890123
  - name: SystemNodeLongSymbol
    node_name: N12345678901234567890123456_0000
    value:
      String: N123456789012345678901234567890123
  - name: SystemNodeLongSymbol
    node_name: N12345678901234567890123456_0001
    value:
      String: N12345678901234567890123456789012
attribute_values_message:
  - name: SystemMessageLongSymbol
    message_id:
      Standard: 9
    value:
      String: SS12345678901234567890123458789012345
  - name: SystemMessageLongSymbol
    message_id:
      Standard: 8
    value:
      String: SS1234567890123456789012345778901
  - name: SystemMessageLongSymbol
    message_id:
      Standard: 7
    value:
      String: SS1234567890123456789012345878901234
  - name: SystemMessageLongSymbol
    message_id:
      Standard: 5
    value:
      String: SS12345678901234567890123456789012
  - name: SystemMessageLongSymbol
    message_id:
      Standard: 4
    value:
      String: S12345678901234567890123456789012
  - name: SystemMessageLongSymbol
    message_id:
      Standard: 0
    value:
      String: M123456789012345678901234567890123
  - name: SystemMessageLongSymbol
    message_id:
      Standard: 2
    value:
      String: M12345678901234567890123456789012
  - name: SystemMessageLongSymbol
    message_id:
      Standard: 3
    value:
      String: MM12345678901234567890123456789012
attribute_values_signal:
  - name: SystemSignalLongSymbol
    message_id:
      Standard: 4
    signal_name: SS123456789012345678901234567890
    value:
      String: SS12345678901234567890123456789012
  - name: SystemSignalLongSymbol
    message_id:
      Standard: 0
    signal_name: SSS12345678901234567890123456789
    value:
      String: SSS12345678901234567890123456789012
  - name: SystemSignalLongSymbol
    message_id:
      Standard: 1
    signal_name: SS1234567890123456789012345_0000
    value:
      String: SS12345678901234567890123456789012332
  - name: SystemSignalLongSymbol
    message_id:
      Standard: 1
    signal_name: SS1234567890123456789012345_0001
    value:
      String: SS1234567890123456789012345678901233
  - name: SystemSignalLongSymbol
    message_id:
      Standard: 1
    signal_name: SS1234567890123456789012345_0002
    value:
      String: SS12345678901234567890123456789012
  - name: SystemSignalLongSymbol
    message_id:
      Standard: 1
    signal_name: S12345678901234567890123456_0000
    value:
      String: S123456789012345678901234567890123
  - name: SystemSignalLongSymbol
    message_id:
      Standard: 2
    signal_name: SS1234567890123456789012345_0003
    value:
      String: SS12345678901234567890123456789012dw
  - name: SystemSignalLongSymbol
    message_id:
      Standard: 2
    signal_name: SS1234567890123456789012345_0004
    value:
      String: SS12345678901234567890123456789012
  - name: SystemSignalLongSymbol
    message_id:
      Standard: 2
    signal_name: S12345678901234567890123456_0001
    value:
      String: S12345678901234567890123456789012
  - name: SystemSignalLongSymbol
    message_id:
      Standard: 2
    signal_name: S12345678901234567890123456_0002
    value:
      String: S123456789012345678901234567890123
  - name: SystemSignalLongSymbol
    message_id:
      Standard: 3
    signal_name: SSS123456789012345678901234_0000
    value:
      String: SSS12345678901234567890123456789012
  - name: SystemSignalLongSymbol
    message_id:
      Standard: 3
    signal_name: SS1234567890123456789012345_0005
    value:
      String: SS1234567890123456789012345678901233
attribute_values_env:
  - name: SystemEnvVarLongSymbol
    variable_name: E12345678901234567890123456_0000
    value:
      String: E12345678901234567890123456789012
signal_groups:
  - message_id:
      Standard: 3
    name: SS1234_sg
    repetitions: 1
    signal_names:
      - SS1234567890123456789012345_0005
      - SSS123456789012345678901234_0000
